### PR TITLE
[TRTLLM-13230][feat] support min_p sampling for TorchSampler

### PR DESCRIPTION
**Pull request by lori-ren**
## Dev Engineer Review

- Added end-to-end `min_p` sampling support to the PyTorch sampler, including validation, request plumbing, strategy resolution, FlashInfer integration, probability renormalization, top-p decay, and speculative-decoding handling.
- `min_p=0` remains a no-op; positive values retain tokens meeting `p >= min_p × p_max`, followed by renormalization. `min_p=1` resolves to greedy decoding.
- Supports composition with temperature, top-k, top-p, and expanded speculative-decoding shapes. One-model speculative decoding explicitly rejects `min_p`.
- Added FlashInfer wrappers for probability-based top-k/top-p sampling and top-k renormalization. Min-p currently falls back to eager execution rather than the CUDA graph path; adding it to graph keys could improve performance but may increase graph-key combinations.
- Follow-up changes correctly apply top-p decay to min-p sampling.
- Validation covers the `[0, 1]` range and updates greedy-resolution behavior consistently across APIs.
- No configuration or test-list files were identified in the provided changes.

## QA Engineer Review

- Added coverage for:
  - Strategy selection and greedy-resolution interactions.
  - `min_p` validation, including disabled and boundary values.
  - Scalar and per-request min-p probability renormalization.
  - Sampling with disabled top-k.
  - Batched sampling and mixed parameter groups.
  - FlashInfer probability handling and retained-token thresholds.
  - Top-p decay dispatch and runtime overrides.
- Modified `test_processed_logprobs_e2e` to explicitly disable min-p processing.
- No test-list coverage mapping was provided for the new or modified test functions; CI/test-list coverage should be confirmed.
- Verdict: needs follow-up.

## Description

The default PyTorch TorchSampler does not apply min_p — the param is accepted by SamplingParams but silently ignored (only the TRT backend / deprecated TRTLLMSampler handle it via the C++ executor SamplingConfig.min_p). Min-P is a common sampling parameter, so this is a feature-parity gap for the PyTorch backend.

Scope:

Apply Min-P in TorchSampler: keep only tokens whose probability ≥ min_p × p_max (where p_max is the highest token probability), then sample from the renormalized distribution.
Wire it into the existing strategy resolution / FlashInfer fast path (resolve_sampling_strategy, sampling_utils_flashinfer.py) — FlashInfer exposes min_p_sampling_from_probs / a min-p renorm op that can be composed with the current top-k/top-p kernels.
Respect existing semantics (SamplingParams definition / C++ runtime default min_p=0.0 = no-op) so behavior matches the legacy backend.
Support combination with temperature / top-k / top-p, and work for speculative-decoding expanded shapes.

Reference: vLLM and SGLang both support Min-P in their GPU samplers.

## Test Coverage

- extended `test_torch_sampler.py` to add tests for min-p sampling

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tests/unittest/_torch/sampler/test_torch_sampler.py [hunk 1]
```diff
@@ -349,6 +351,79 @@ def test_top_k_top_p_with_temperature(self):
...
+    def test_min_p_with_top_k_top_p(self):
+        params = SamplingParams(min_p=0.1, top_k=42, top_p=0.7, temperature=0.8)
+        request = self._build_mock_llm_request(params)
+        strat = _request_strategy(request, vocab_size=self.VOCAB_SIZE)
+        assert len(strat) == 5
+        assert strat[0] == "min_p"
+        assert strat[1] == 42
+        assert strat[2] == pytest.approx(0.7)
+        assert strat[3] == pytest.approx(0.1)
+        assert strat[4] == pytest.approx(0.8)
```
[the remark this answers is not in the document]
Comment (lori-ren) [reply tests/unittest/_torch/sampler/test_torch_sampler.py:409]: The check here is only for matching other tests for parameters. Mixed strategies are covered via TestBatchedSampling, not here.

File: tensorrt_llm/_torch/pyexecutor/sampler/ops/vanilla.py [hunk 2]
```diff
@@ -50,19 +50,37 @@ class BeamSearchMetadata(StrategyMetadata):
...
+
+    ``min_p`` is a scalar or a per-request tensor.
+    """
+    max_probs = probs.max(dim=-1, keepdim=True).values
+    if isinstance(min_p, torch.Tensor):
+        min_p = min_p.reshape(-1, 1)
+    thresholds = min_p * max_probs
+    probs = torch.where(probs < thresholds, torch.zeros_like(probs), probs)
+    probs = probs / probs.sum(dim=-1, keepdim=True)
+    return probs
```
[the remark this answers is not in the document]
Comment (mikeiovine) [reply tensorrt_llm/_torch/pyexecutor/sampler/ops/vanilla.py:68]: Let's adopted the proposed fix

File: tensorrt_llm/_torch/pyexecutor/sampler/sampler_strategy.py [hunk 3]
```diff
@@ -154,6 +163,12 @@ def resolve_sampling_strategy(params: UtilsSamplingParams, *, vocab_size: int) -
     assert top_p > 0, "non-greedy sampling requires valid top_p"
     need_top_p = top_p < 1
 
+    # Disabled top_k is 0 ("keep all"), not vocab_size, which can be the
+    # fast-greedy probe (2**31) and overflow the int32 tensor; _compute_probs
+    # sanitizes it.
+    if min_p is not None and min_p > 0.0:
+        return ("min_p", top_k if need_top_k else 0, top_p, min_p, temperature)
+
```
[the remark this answers is not in the document]
Comment (mikeiovine) [reply tensorrt_llm/_torch/pyexecutor/sampler/sampler_strategy.py:196]: Let's apply a fix of this comment too. In general we should make sure that this works with speculative decoding, both rejection sampling and exact match verification

**Review by mikeiovine — APPROVED** [review]
According to coderabbit, the one engine spec dec path is currently ignoring min p. That's fine for now, but we should document the limitation and fix it in a follow up.

We should fix that as soon as possible. We've had issues with spec dec ignoring sampling params in the past, users find it confusing

**Comment by lori-ren** [comment]
Hi @mikeiovine, I've extended min_p support for one-eng spec path. Pls review at your convenience.

**Review by zhaoyangwang-nvidia — CHANGES_REQUESTED** [review]
In `sample_and_accept_draft_tokens `(the target-probs path) you correctly short-circuit via skip_min_p, passing None so the if min_p is not None branch in compute_probs_from_logits gets traced away and no min-p op enters the graph:


interface.py ~L1732
```
min_ps = (None if spec_metadata.skip_min_p else spec_metadata.min_ps[gen_start:gen_end])
But the draft-sampling call sites pass the tensor unconditionally instead of honoring skip_min_p:

```

interface.py ~L1552, ~L1580, ~L1960, ~L1973, ~L2246, ~L2259
```
min_ps = spec_metadata.request_min_p[:batch_size]     # never None

sampling_batch_spec_dec_one_model(..., min_p=min_ps, ...)
```
Because request_min_p / min_ps is a torch.zeros(...) buffer slice (never None), the if min_p is not None guard in compute_probs_from_logits is always True at torch.compile/CUDA-graph capture time. This bakes min_p_renorm_probs into the captured graph and runs it on every replay, including for the very common case where no request enables min-p.

min_p_renorm_probs is a numerical no-op when min_p == 0 (thresholds become 0, where keeps everything, sum→div by 1.0), but it is not free: it adds a full-vocab row-max reduction + elementwise where + row-sum + divide per draft step, per replay. On the draft path this fires once per draft token, so the wasted work scales with draft length.

And please test the spec decoding performance with https://gitlab-master.nvidia.com/ftp/spec-benchmarks

**Comment by zhaoyangwang-nvidia** [comment]
Please take a look of https://github.com/NVIDIA/TensorRT-LLM/pull/16561 we need add a skip min-p filter.

**Comment by lori-ren** [comment]
> Please take a look of #16561 we need add a skip min-p filter.

Done, now min_p fallbacks to eager execution while top_p/top_k/greedy remain the same CUDA graph path. We may add min_p to graph key to improve min_p performance, but this could lead to a combinatorial explosion.

**Comment by lori-ren** [comment]
@zhaoyangwang-nvidia pls help review this PR

File: tensorrt_llm/_torch/pyexecutor/sampler/ops/vanilla.py [hunk 4]
```diff
@@ -117,6 +137,9 @@ def top_k_top_p_sampling_batch(
     else:
         probs = torch.softmax(logits, dim=-1)
 
+    if min_p > 0:
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/pyexecutor/sampler/ops/vanilla.py]: The min_p application order here contradicts the flashinfer path.  please check.

Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/pyexecutor/sampler/ops/vanilla.py]: _StrategyImpls._compute_probs (sampling_utils.py:402-410) applies min_p first

File: tensorrt_llm/_torch/pyexecutor/sampler/ops/vanilla.py [hunk 5]
```diff
@@ -50,26 +50,46 @@ class BeamSearchMetadata(StrategyMetadata):
...
+) -> torch.Tensor:
+    """Keep tokens with prob >= ``min_p`` times the per-row max, then renormalize.
+
+    ``min_p`` is a scalar or a per-request tensor.
+    """
+    max_probs = probs.max(dim=-1, keepdim=True).values
+    if isinstance(min_p, torch.Tensor):
+        min_p = min_p.reshape(-1, 1)
+    thresholds = min_p * max_probs
+    probs = torch.where(probs < thresholds, torch.zeros_like(probs), probs)
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/pyexecutor/sampler/ops/vanilla.py]: Could we use `probs = probs.masked_fill(probs < thresholds, 0.0)` ?

**Review by zhaoyangwang-nvidia — APPROVED** [review]
Please update the doc file docs/source/features/sampling.md

File: docs/source/features/sampling.md [hunk 6]
```diff
@@ -115,9 +119,8 @@ llm.generate(["Hello, my name is",
   * Top-P decay is not supported in combination with beam search or with speculative decoding
     modes that route draft tokens through the Torch Sampler; such requests are rejected.
 
-* If `no_repeat_ngram_size = n` is specified, any token that would recreate an `n`-gram already
-  present in the sequence (prompt included) is excluded from sampling. `None` or `0` disables
-  the restriction.
+* Min-P is not supported in combination with one-model speculative decoding. Such requests are
+  rejected at admission.
```
[the remark this answers is not in the document]
Comment (lori-ren) [reply docs/source/features/sampling.md]: Done, changed that statement to "Positive min_p ..."

**Review by QiJune — APPROVED** [review]
